pkgs/by-name/zl/zluda/package.nix

```diff
@@ -0,0 +1,79 @@
+{ lib, fetchFromGitHub, rocmPackages, python3, cargo, rustc, cmake, clang, zlib, libxml2, libedit, rustPlatform, stdenv }:
+
+rustPlatform.buildRustPackage rec {
+  pname = "zluda";
+  version = "3";
+
+  src = fetchFromGitHub {
+    owner = "vosen";
+    repo = "ZLUDA";
+    rev = "v${version}";
+    hash = "sha256-lykM18Ml1eeLMj/y6uPk34QOeh7Y59i1Y0Nr118Manw=";
+    fetchSubmodules = true;
+  };
+
+  cargoPatches = [ ./ZLUDA-cargo.lock.patch ];
+
+  buildInputs = [
+    rocmPackages.clr
+    rocmPackages.miopen
+    rocmPackages.rocm-smi
+    rocmPackages.rocsparse
+    rocmPackages.rocsolver
+    rocmPackages.rocblas
+    rocmPackages.hipblas
+    rocmPackages.rocm-cmake
+    rocmPackages.hipfft
+  ];
+
+  nativeBuildInputs = [
+    python3
+    cargo
+    rustc
+    cmake
+    clang
+    zlib
+    libxml2
+    libedit
+  ];
+
+  preConfigure = ''
+    # Comment out zluda_blaslt in Cargo.toml
+    sed -i '/zluda_blaslt/d' Cargo.toml
+    # TODO: investigate test failure (the test seems to require build time env vars that aren't set on linux?)
+    rm zluda_inject/tests/inject.rs
+  '';
+
+  buildPhase = ''
+    runHook preBuild
+    cargo xtask --release
+    runHook postBuild
+  '';
+  # xtask doesn't support passing --target, but nix hooks expect the folder structure from when it's set
+  env.CARGO_BUILD_TARGET = stdenv.hostPlatform.rust.cargoShortTarget;
+
+  cargoHash = "sha256-gZdLThmaeWVJXoeG7fuusfacgH2RNTHrqm8W0kqkqOY=";
+
+  # vergen panics if the .git directory isn't present
+  # Disable vergen and manually set env
+  postPatch = ''
+    substituteInPlace zluda/build.rs \
+      --replace-fail 'vergen(Config::default())' 'Some(())'
+  '';
+  env.VERGEN_GIT_SHA = "1b9ba2b2333746c5e2b05a2bf24fa6ec3828dcdf-nix";
+
+  preInstall = ''
+    mkdir -p $out/lib/
+    find target/release/ -maxdepth 1 -type l -name '*.so*' -exec \
+      cp --recursive --no-clobber --target-directory=$out/lib/ {} +
```

Comment:
(not a change request)

- Do buildRustPackage/cargo setup hooks not install these automatically?
- Do we need `--recursive` for *.so*`?..

Reply:
* `buildRustPackage` doesn't install these automatically because they're created by an upstream script that doesn't respect cargo's environment variables correctly (and places them into `target/release` instead of `target/$ARCH/release`.)
* I think you need `--recursive` here to correctly copy the symlinks instead of turning them into files. But my memory is hazy.